tests/bsim/bluetooth/host/att/long_read/main.c

```diff
@@ -0,0 +1,220 @@
+/* Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <argparse.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/settings/settings.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/bluetooth/bluetooth.h>
+
+#include "testlib/adv.h"
+#include "testlib/att_read.h"
+#include "testlib/att_write.h"
+#include "testlib/bs_macro.h"
+#include "testlib/bs_sync.h"
+#include "testlib/conn_ref.h"
+#include "testlib/conn_wait.h"
+#include "testlib/connect.h"
+#include "testlib/log_utils.h"
+#include "testlib/scan.h"
+#include "testlib/security.h"
+
+LOG_MODULE_REGISTER(main, LOG_LEVEL_DBG);
+
+#define UUID_1                                                                                     \
+	BT_UUID_DECLARE_128(0xdb, 0x1f, 0xe2, 0x52, 0xf3, 0xc6, 0x43, 0x66, 0xb3, 0x92, 0x5d,      \
+			    0xc6, 0xe7, 0xc9, 0x59, 0x9d)
+
+#define UUID_2                                                                                     \
+	BT_UUID_DECLARE_128(0x3f, 0xa4, 0x7f, 0x44, 0x2e, 0x2a, 0x43, 0x05, 0xab, 0x38, 0x07,      \
+			    0x8d, 0x16, 0xbf, 0x99, 0xf1)
+
+static ssize_t read_mtu_validation_chrc(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+					void *buf, uint16_t buf_len, uint16_t offset)
+{
+	ssize_t read_len;
+
+	LOG_INF("Server side buf_len %u", buf_len);
+
+	/* Note: We assume `buf_len` is equal to the usable payload
+	 * capacity of the response PDU. I.e. `(ATT_MTU - 1)` for
+	 * BT_ATT_OP_READ_RSP and BT_ATT_OP_READ_BLOB_RSP.
+	 */
+
+	/* Send back a full PDU on the first read (on offset 0). Then an
+	 * not full one for the second read to conlude the long read..
+	 */
+	read_len = buf_len;
+	if (offset > 0) {
+		__ASSERT_NO_MSG(read_len > 0);
+		/* The second PDU is one-less-than-full to test for off
+		 * by one errors.
+		 */
+		read_len -= 1;
+	}
+
+	/* If the ATT_MTU is too large, sending a one-less-than-full
+	 * response would exeed the max attribute length limit.
+	 */
+	__ASSERT(buf_len < (BT_ATT_MAX_ATTRIBUTE_LEN / 2),
+		 "The EATT buffer is too large for this test.");
+
+	/* Fill the read buffer with zero. */
+	memset(buf, 0, read_len);
```

Comment:
I don't think it makes any practical difference, right? Semantically I'm preparing a `read_len` number of bytes response.

Reply:
True, but the code does not match the comment: `/* Fill the read buffer with zero. */`. 

Since you are later overriding `read_len` bytes, then this expression doesn't really do anything, right?

Reply:
Fixed